extide/o.apache.tools.ant.module/src-bridge/org/apache/tools/ant/module/bridge/impl/ForkedJavaOverride.java

```diff
@@ -162,219 +166,216 @@ public synchronized void setError(File[] error) {
-                Vector v = getProject().getBuildListeners();
-                for (Object o : v) {
-                    if (o instanceof NbBuildLogger) {
-                        NbBuildLogger l = (NbBuildLogger) o;
-                        err = logLevel != Project.MSG_INFO;
-                        ow = err ? l.err : l.out;
-                        session = l.thisSession;
-                        break;
-                    }
-                }
-            }
             try {
+                byte[] data = new byte[1024];
+                int len;
+
                 try {
-                    int c;
-                    while ((c = in.read()) != -1) {
-                        if (logLevel == null) {
-                            // Input gets sent immediately.
-                            out.write(c);
-                            out.flush();
-                        } else {
-                            synchronized (this) {
-                                if (c == '\n') {                                    
-                                    String str = currentLine.toString(encoding);
-                                    int len = str.length();
-                                    if (len > 0 && str.charAt(len - 1) == '\r') {
-                                        str = str.substring(0, len - 1);
-                                    }
+                    while ((len = in.read(data)) >= 0) {
+                        int last = 0;
+                        for (int i = 0; i < len; i++) {
+                            int c = data[i] & 0xff;
 
+                            // Add folds for stack traces and mark up lines
+                            // not processed by JavaAntLogger stack trace detection
+                            if (c == '\n') {
+                                String str = appendAndTake(data, last, i > last && data[i - 1] == '\r' ? i - last - 1 : i - last);
+
+                                synchronized (foldingHelper) {
                                     foldingHelper.checkFolds(str, err, session);
-                                    if (str.length() < LOGGER_MAX_LINE_LENGTH) { // not too long message, probably interesting
-                                        // skip stack traces (hyperlinks are created by JavaAntLogger), everything else write directly
-                                        if (!STACK_TRACE.matcher(str).find()) {
-                                            StandardLogger.findHyperlink(str, session, null).println(session, err);
-                                        }
-                                    } else {
-                                        // do not match long strings, directly create a trivial hyperlink
+                                    if (str.length() >= LOGGER_MAX_LINE_LENGTH || !STACK_TRACE.matcher(str).matches())
                                         StandardLogger.findHyperlink(str, session, null).println(session, err);
```

Comment:
gonna have to check if matches() gives the same results as find() with the given regex.

Reply:
I'm pretty certain it does because it contains .* on both ends so will always match the whole string anyway.  Also probably more to the point this is just to check if it should ignore the lines because JavaAntLogger will mark them up - and JavaAntLogger uses the same regex and uses matches().

Reply:
yes I think you are right. The regex already matches the full line, `find()` would be redundant and likely much slower.
I mostly tried to find an excuse to use the regex checker window #4200.

Reply:
For the test i was using, find vs matches is roughly ~1500x slower and there are worse cases.  Due to the greedy operator it matches the whole string then walks all the way back to the beginning testing each position looking for the next substring match.  With matches() it then terminates but because it isn't anchored to the start of the string find() advances by one character and repeats - all the way to the other end - so it's particularly bad with long lines, and is the reason for the line length checking here and in javaantlogger. I filed a jdk performance issue against java.util.regex.Pattern but my initial assessment was wrong as to the cause, and it would be tricky to address due to the way the state is managed in the compiled expression.

perl doesn't even execute the regex because it does a pre-scan and determines the string can't possibly match due to the bits that aren't optional which are not present like '('.

Removing the .* at either end makes find() run about the same as matches().  With some simplification and no capturing groups it can run somewhat faster than calling matches() but it's no longer a significant part of the processing time and doesn't seem worth maintaining another equally unreadable expression.